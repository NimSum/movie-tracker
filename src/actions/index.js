export const addMovies = (movies) => ({
  type: 'ADD_MOVIES',
  movies
})

export const formType = (type) => ({
  type: 'FORM_TYPE',
  form: type
})

export const currentUser = (id, name, email, favorites) => ({
  type: 'CURRENT_USER',
  userInfo: { id, name, email, favorites }
})

export const selectedMovie = (movie) => ({
  type: 'SELECTED_MOVIE',
  movie
})

export const updateFavorites = (favorite) => ({
  type: 'UPDATE_FAVORITES',
  favorite
})