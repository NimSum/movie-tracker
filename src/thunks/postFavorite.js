import { postUser } from '../utils/apiFetches/postUser.js'

export const postFavorite = (body) => {
  return async (dispatch) => {
    try {
      const url = 'http://localhost:3000/api/users/favorites/new'
      await postUser(url, body)
    } catch(error) {
      throw new Error(error, 'Failed to post favorite')
    }
  }

}