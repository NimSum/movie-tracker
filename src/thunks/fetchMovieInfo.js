import { fetchAnything } from '../utils/apiFetches/fetchAnything.js'
import{ selectedMovie } from '../actions/index'

export const fetchMovieInfo = (url) => {
  return async (dispatch) => {
    try {
      const movieInfo = await fetchAnything(url)
      dispatch(selectedMovie(movieInfo))
    } catch(error) {
      throw new Error(error, 'Problem getting movie info :(')
    }
  }
}