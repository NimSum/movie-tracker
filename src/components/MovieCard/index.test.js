import React from 'react';
import { shallow } from 'enzyme';
import { MovieCard, mapStateToProps, mapDispatchToProps } from './index';
import { fetchMovieInfo } from '../../thunks/fetchMovieInfo';
import { updateFavorites } from '../../actions'

jest.mock('../../thunks/fetchMovieInfo');
jest.mock('../../actions/');

describe('MovieCard', () => {
  const mockFetchInfo = jest.fn();
  const mockUpdateFavorites = jest.fn();
  const mockMovies = [
    {title: 'Avengers Infinity War'}
  ]
  const mockMovie = {
    movie_id: 111,
    title: 'Avengers End Game',
    poster_path: 'poster.jpg',
    release_date: '2019-93-19',
    vote_average: '55.5',
    overview: 'Thanos dies'
  }
  let wrapper;
  let mockDispatch;
  beforeEach(() => {
    wrapper = shallow( 
      < MovieCard 
        card={ mockMovie }
        fetchInfo={ fetchMovieInfo }
        updateFavorites = { mockUpdateFavorites }
        userId={ 1 }
        userFavorites= { updateFavorites } />
    )
  })

  afterEach(() => {
    mockDispatch.mockClear();
    fetchMovieInfo.mockClear();
  })

  it('should have initial state', () => {
    const expected = {
      cardHover: false,
      loginPrompt: false
    }
    expect(wrapper.state()).toEqual(expected)
  })

  it('should match component snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  })

  it('should fetch showmore info', () => {
    wrapper.instance().showMoreInfo();
    expect(fetchMovieInfo).toHaveBeenCalledTimes(1);
  })

  it('should toggle favorite', () => {
    wrapper.instance().toggleFavorite();
    expect(mockUpdateFavorites).toHaveBeenCalledTimes(1)
  })

  it('should set state loginprompt false if user is valid', () => {
    wrapper.setState({ loginPrompt: true })
    wrapper.instance().toggleFavorite();
    expect(wrapper.state().loginPrompt).toEqual(false);
  })

  it('should toggle loginPrompt', () => {
    wrapper = shallow( 
      < MovieCard 
        card={ mockMovie }
        userFavorites= { updateFavorites } />
    )
    wrapper.instance().toggleFavorite();
    expect(wrapper.state().loginPrompt).toEqual(true);
    wrapper.instance().toggleFavorite();
    expect(wrapper.state().loginPrompt).toEqual(false);
  })

  it('should invoke showmoreinfo on click', () => {
    wrapper.find('.more-info-btn').simulate('click')
    expect(fetchMovieInfo).toHaveBeenCalledTimes(1)
  })

  it('should invoke toggleFavorite on click', () => {
    wrapper.setState({ loginPrompt: true })
    wrapper.find('.favorite-btn').simulate('click');
    expect(wrapper.state().loginPrompt).toEqual(false);
  })

  describe('mapStateToProps', () => {
    const mockState = {
      currentUser: { 
        id: 1,
        favorites: [
          {title: 'Avengers Infinity War'}
        ]
      }
    }
    const expected = {
      userId: 1,
      userFavorites: [
        {title: 'Avengers Infinity War'}
      ]
    }
    const mappedProps = mapStateToProps(mockState);
    expect(mappedProps).toEqual(expected);
  })

  describe('mapDispatchToProps', () => {
    mockDispatch = jest.fn();
    const mockUrl = 'www.nimsum.com/url/url'
    const mockMovie = [
      {title: 'Avengers Infinity War'}
    ]
    const dispatchUpdateFav = updateFavorites(mockMovie);
    const dispatchFetchInfo = fetchMovieInfo(mockUrl);
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.fetchInfo(mockUrl);
    expect(mockDispatch).toHaveBeenCalledWith(dispatchUpdateFav);
    mappedProps.updateFavorites(mockMovie);
    expect(mockDispatch).toHaveBeenCalledWith(dispatchFetchInfo)
  })
})