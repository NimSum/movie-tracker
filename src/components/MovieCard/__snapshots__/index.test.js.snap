// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard should match component snapshot 1`] = `
<article
  className="movie-card"
  key="111"
>
  <img
    alt="Promotional movie poster for Avengers End Game"
    className="poster-img"
    src="https://image.tmdb.org/t/p/w500/poster.jpg"
  />
  <div
    className="card-hover"
  >
    <h4
      className="card-hover-heading"
    >
      Avengers End Game
    </h4>
    <Link
      to="/movies/111"
    >
      <button
        className="more-info-btn btn"
        onClick={[Function]}
      >
        More 
        <br />
         Info
      </button>
    </Link>
    <button
      className="favorite-btn btn"
      onClick={[Function]}
    >
      <img
        alt="Star icon"
        className="inactive-star star"
        src="inactive.svg"
      />
    </button>
    <p
      className="prompt-placeholder prompt"
    >
      X
    </p>
  </div>
</article>
`;
